pkg/controllers/state/cluster.go

```diff
@@ -96,25 +98,31 @@ func (c *Cluster) Synced(ctx context.Context) bool {
 		return false
 	}
 	c.mu.RLock()
-	stateNames := sets.New(lo.Keys(c.nameToProviderID)...)
+	stateMachineNames := sets.New[string]()
+	stateNodeClaimNames := sets.New[string]()
+	for k := range c.nodeClaimKeyToProviderID {
+		if k.IsMachine {
+			stateMachineNames.Insert(k.Name)
+		} else {
+			stateNodeClaimNames.Insert(k.Name)
+		}
+	}
+	stateNodeNames := sets.New(lo.Keys(c.nodeNameToProviderID)...)
 	c.mu.RUnlock()
 
-	names := sets.New[string]()
+	machineNames := sets.New[string]()
 	for _, machine := range machineList.Items {
-		// If the machine hasn't resolved its provider id, then it hasn't resolved its status
-		if machine.Status.ProviderID == "" {
```

Comment:
Don't we need this? Since this function does a list call on the active machines/nodes, this should ensure that we wait for any machines/nodes not in cluster state that don't have a providerID, and consider the cluster state not synced.

Reply:
Ah that's fair, but wouldn't keeping this in here short circuit all the computation we do here anyway? If we know a machine without a providerRef is going to make L125 fail anyway, seems like we can just return false here anyway.

Reply:
It was more about making simplifying assumptions to not overcomplicate. I could go either way. It's technically redundant, but I'm fine keeping it if you feel like we should

Reply:
I think it's more clear to the reader that this is a way that this can fail, rather than having to infer and debug later.

Reply:
Updated